File: blue-core/src/blue/noteProcessor/RotateProcessor.java
package blue.noteProcessor;

import blue.BlueSystem;
import blue.soundObject.Note;
import blue.soundObject.NoteList;
import blue.soundObject.NoteParseException;
import blue.utility.ScoreUtilities;
import electric.xml.Element;
import java.util.Collections;

/**
 * Title: blue Description: an object composition environment for csound
 * Copyright: Copyright (c) 2001 Company: steven yi music
 * 
 * @author steven yi
 * @version 1.0
 */

public class RotateProcessor implements NoteProcessor, java.io.Serializable {

    int noteIndex = 1;

    public RotateProcessor() {
    }

    public String toString() {
        // return "[sublist] start: " + start + " end: " + end;
        return "[rotate]";
    }

    public String getNoteIndex() {
        return Integer.toString(noteIndex);
    }

    public void setNoteIndex(String noteIndex) {
        this.noteIndex = Integer.parseInt(noteIndex);
    }

    public final void processNotes(NoteList in) throws NoteProcessorException {
        if (in.size() < 2 || noteIndex == 1) {
            return;
        }

        in.sort();

        Note lastNote = in.getNote(in.size() - 1);

        float startTime = lastNote.getStartTime()
                + lastNote.getSubjectiveDuration();

        int index = noteIndex;
        if (index > 0) {
            index = index - 1;
        } else {
            index = in.size() + index;
        }

        if (index > in.size()) {
            throw new NoteProcessorException(this, BlueSystem
                    .getString("noteProcessorException.rotateIndex"));
        }
        Collections.rotate(in, -index);

        index = in.size() - index;

        while (index < in.size()) {
            Note n = in.getNote(index);
            n.setStartTime(n.getStartTime() + startTime);
            index++;
        }

        ScoreUtilities.normalizeNoteList(in);

    }

    public static void main(String args[]) {
        NoteList n = new NoteList();

        for (int i = 0; i < 10; i++) {
            try {
                n.addNote(Note.createNote("i1 " + (i * 2) + " 5 " + i + " 4"));
            } catch (NoteParseException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }

        System.out.println("before: \n\n" + n + "\n\n");

        RotateProcessor rlp = new RotateProcessor();
        rlp.setNoteIndex("-3");
        try {
            rlp.processNotes(n);
        } catch (NoteProcessorException ex) {
            System.out.println("Exception: " + ex.getMessage());
        }

        System.out.println("after: \n\n" + n + "\n\n");
    }

    public static NoteProcessor loadFromXML(Element data) {
        RotateProcessor rlp = new RotateProcessor();

        rlp.setNoteIndex(data.getElement("noteIndex").getTextString());

        return rlp;
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.noteProcessor.NoteProcessor#saveAsXML()
     */
    public Element saveAsXML() {
        Element retVal = new Element("noteProcessor");
        retVal.setAttribute("type", this.getClass().getName());

        retVal.addElement("noteIndex").setText(this.getNoteIndex());

        return retVal;
    }
}

File: packages/blue-data/src/note-processors/rotate-processor.ts
import { NoteProcessor } from './note-processor';
import { NoteProcessorException } from './note-processor-exception';
import { NoteList } from '../sound-objects/note-list';
import { Element } from '../serialization/xml-reader';
import { normalizeNoteList } from '../utilities/score';

const JAVA_TYPE = 'blue.noteProcessor.RotateProcessor';

export class RotateProcessor extends NoteProcessor {
  private _noteIndex = 1;

  constructor();
  constructor(src: RotateProcessor);
  constructor(src?: RotateProcessor) {
    super();
    if (src) {
      this._noteIndex = src._noteIndex;
    }
  }

  getNoteIndex(): string {
    return this._noteIndex.toString();
  }
  setNoteIndex(noteIndex: string): void {
    this._noteIndex = parseInt(noteIndex, 10);
  }

  override process(notes: NoteList): NoteList {
    if (notes.length < 2 || this._noteIndex === 1) {
      return notes;
    }

    notes.sort();

    const lastNote = notes.getNote(notes.length - 1);
    const startTime = lastNote.getStartTime() + lastNote.getSubjectiveDuration();

    let index = this._noteIndex;
    if (index > 0) {
      index = index - 1;
    } else {
      index = notes.length + index;
    }

    if (index > notes.length) {
      throw new NoteProcessorException('Rotate index out of bounds', index);
    }

    const rotated = new NoteList();
    for (let i = index; i < notes.length; i++) {
      rotated.push(notes.getNote(i));
    }
    for (let i = 0; i < index; i++) {
      rotated.push(notes.getNote(i));
    }

    let idx = notes.length - index;
    while (idx < notes.length) {
      const n = rotated.getNote(idx);
      n.setStartTime(n.getStartTime() + startTime);
      idx++;
    }

    notes.clear();
    notes.merge(rotated);

    normalizeNoteList(notes);
    return notes;
  }

  override getDisplayName(): string {
    return 'RotateProcessor';
  }

  override deepCopy(): RotateProcessor {
    return new RotateProcessor(this);
  }

  saveAsXML(): Element {
    const elem = new Element('noteProcessor');
    elem.setAttribute('type', JAVA_TYPE);
    elem.addElement('noteIndex').setText(this.getNoteIndex());
    return elem;
  }

  static loadFromXML(data: Element): RotateProcessor {
    const proc = new RotateProcessor();
    const ni = data.getTextString('noteIndex');
    if (ni !== null) proc._noteIndex = parseInt(ni, 10);
    return proc;
  }
}
